### ISSUE TITLE
`swipeViewWithAccessibilityLabelL:inDirection:` does not wait for app to idle

### ISSUE THREAD

**Issue by nathanbeaumont** [kif-framework/KIF#1090]
We are seeing this on iOS 12, Xcode 10. Seeing this on multiple simulators iPhone 7, XS, etc.

When calling `swipeViewWithAccessibilityLabel: inDirection:` the application does not wait for the scroll to finish.  This is causing tests in tableviews to fail, because the tester selects a tableview cell while the tableview is still scrolling.

**Comment by stevencuy** [comment]
For some clarification, I believe this is specific to iOS 12, as this issue does not occur when running iOS 11 simulators with Xcode 10

**Comment by justinseanmartin** [comment]
I was out on vacation last week, so just getting back to this. Speaking of which, if you're using KIF regularly and want to help maintain it, I'm looking for more folks that are interested in helping contribute to the project.

Do you want to give the branch [justinseanmartin:jmartin/issue-1090](https://github.com/justinseanmartin/KIF/tree/jmartin/issue-1090) a try and see if this helps? I based it off of latest master, but should also be a trivial patch to apply to whatever version of KIF you're currently using.

**Comment by justinseanmartin** [comment]
@stevencuy - Can you elaborate? What is the behavior when they fail? The only thing I changed was to wait for the animations to complete before it yields back to the tests to perform the next action (presumably the tap action).

**Comment by stevencuy** [comment]
The same interaction of 

- Successfully swipes view
- Fails to tap view

But fails in tests that were previously not failing. Some of these tests are very simple containing only 2 waits prior to the interaction above.

**Comment by stevencuy** [comment]
The scroll animation is still going when the tap fails. The only fix at the moment is having flat waits `waitForTimeInterval:` following the swipe. `waitForAnimationsToFinish:` does not fix the issue. The element is visible in the simulator when the tap gesture is called, and does not fail to find the element or tap it. But after executing the tap, the view is not actually tapped and the next step of the test fails to find the element not displayed. 

I'll debug further once I have availability and report back.

**Comment by justinseanmartin** [comment]
If the view is still scrolling when the tap touch event is being sent, it likely is being dropped due to user interaction being disabled. 

I guess the scroll `dragPointsAlongPaths:` action isn't really an animation as far as UIKit is concerned, which maybe makes sense (though I'd have thought the bounds property was being animated). You could double check this by seeing if during the scroll action if any of the scrollviews `[view.layer hasAnimations]` reports true.

If that's the case, then one thought might be to check all scroll views and make sure that their bounds have stopped changing before returning from `dragPointsAlongPaths:`.

**Comment by justinseanmartin** [comment]
I just pushed to my branch again with a sloppy attempt at the previous suggestion of checking the frame/bounds. If this fixes your issues, I'll clean it up and open a PR.

**Comment by billburgess** [comment]
Any update on this? I'm having failures with this method. In my case it only happens when using CircleCI and the swipe action doesn't seem to trigger correctly causing the tap actions behind it to fail. I have not been able to reproduce this locally. If anyone has any workarounds or ideas I can try I'd appreciate it. I've wasted hours on this problem so far in CI minutes and no fix in sight.

**Comment by justinseanmartin** [comment]
Did you try targetting my branch with the changes that I'd made 2 weeks ago to try and mitigate the issue to see if that worked?

You could try enabling `KIF_PRINTVIEWTREEONFAILURE` and see if that yields more information about what is onscreen when the failure happens. I'm not sure if CircleCI has support for attaching screenshots to test failures or anything like that, but I've also used `KIF_SCREENSHOTS` to give a more visual cue as to the state of the world when the test fails.

Is it failing consistently or sporadically for you in CircleCI? If you add a `[viewTester waitForTimeInterval:10.0]` before attempting the swipe action, does that help? I'm not suggesting this as a permanent fix, but as a debugging measure to understand if the issue is that we're not waiting long enough or something else.

Just to check, you're running with the same Xcode #/iOS #/MacOS #/Sim Device combination as your CI job is configured to use?

**Comment by billburgess** [comment]
I wasn't using that long of a delay. Is 10 seconds really needed to make that work? I will give the other branch a try to see if that help.

**Comment by justinseanmartin** [comment]
Does tapping and swiping the switch do the same thing? We had issues with our switches as well for some reason (I can't remember exactly why at this point), but our workaround was to add an extension on KIFUIViewTestActor that would do:

```
@implementation KIFUIViewTestActor (SQSwitch)

- (void)setSQUISwitchOn:(BOOL)switchIsOn;
{
    @autoreleasepool {
        [viewTester waitForAnimationsToFinish];

        SQUISwitch *foundSwitch = (SQUISwitch *)[[self usingPredicate:[NSPredicate screenPredicateWithClass:SQUISwitch.class]] waitForView];

        if (foundSwitch.isOn != switchIsOn) {
            [self tap];
        }
    }
}

@end
```

Also, the `screenPredicateWithClass` method is:

```
@implementation NSPredicate (ScreenPredicateAdditions)

+ (instancetype)screenPredicateWithClass:(Class)cls;
{
    return [NSPredicate predicateWithBlock:^BOOL(id evaluatedObject, NSDictionary *bindings) {
        return [evaluatedObject isKindOfClass:cls];
    }];
}

@end
```

IIRC - The issue we had was the tap event was being sent to the center table view cell containing the switch which did nothing.

**Comment by billburgess** [comment]
@justinseanmartin The issue only happens when running on CircleCI for me. I can't reproduce it locally at all.

**Comment by billburgess** [comment]
I've updated to Xcode 10.1 using 12.1 simulators. I'm also using @justinseanmartin PR branch and the issue is still happening. All of my 10.3.1 tests work just fine. Anything running 12 is busted completely. As it pertains to swiping against a UISwitch anyway.

**Comment by justinseanmartin** [comment]
@billburgess - Please file a separate issue. I think it is unrelated to the issue originally called out, which was related to table views specifically. If you can add a test to the functional test suite in KIF, I'd be happy to take a look to see if I can help figure out the best way to fix it.

**Comment by justinseanmartin** [comment]
Closing, as I haven't heard back from @nathanbeaumont and @stevencuy on the original issue or if my branch helps resolve the issue. Please feel free to reopen if there is more information available about the table view issue on using `swipeViewWithAccessibilityLabel:`.
